ui/app/AppLayouts/stores/RootStore.qml

```diff
@@ -13,6 +13,7 @@ QtObject {
     property var aboutModuleInst: aboutModule
     property var communitiesModuleInst: communitiesModule
     property var observedCommunity: communitiesModuleInst.observedCommunity
+    property var globalUtilsInst: typeof globalUtils !== "undefined" ? globalUtils : null
```

Comment:
@micieslak if course in the future the idea is to get rid of context properties (currently are there because of nim) and have QML types instead. At least that's the info I have of how the Qt/C++ layer will be implemented (@stefandunca @saledjenic please correct me if I'm wrong). That said, all these `xxxInst` properties are there to prepare for the upcoming QML types, something like [here](https://codereview.qt-project.org/gitweb?p=qt-apps/neptune3-ui.git;a=blob;f=apps/com.pelagicore.music/stores/MusicStore.qml;h=04f9fd244d5e11635c2ce0f00b1d222f80cf56ee;hb=7f3bda0bdc9c4c1c704fbcf8585e3470f9f9e9f5) for example if you see
```
 property FilterAndBrowseModel searchAndBrowseModel: FilterAndBrowseModel {
           id: searchBrowseModel
```

Considering the above, I wouldn't change anything in stores at the moment until we have the Qt layer in.

Reply:
That's true that new qt/cpp layer will bring chances to fix many problems in the future, but for now if we want to improve separation of concerns and testability we can hide access to context properties in stores instead of exposing them (keep it as an implementation detail of stores). Than we will gain clear separation between UI and backend parts defined by stores' API.

For me it's hard to refer to [here](https://codereview.qt-project.org/gitweb?p=qt-apps/neptune3-ui.git;a=blob;f=apps/com.pelagicore.music/stores/MusicStore.qml;h=04f9fd244d5e11635c2ce0f00b1d222f80cf56ee;hb=7f3bda0bdc9c4c1c704fbcf8585e3470f9f9e9f5) in context of our project because it's sample app far from real project, it uses automotive-specific backend system (https://doc.qt.io/QtInterfaceFramework/backend-system.html) which is not present and not needed in our case and additionally the code is quite messy (many properties probably should be marked as readonly) and contains a lot of js code.

For clarification and to better explain what I mean I'm referring to https://doc.qt.io/Neptune3UI/neptune3ui-application-architecture.html#stores:

> A store is a data-driven object, that encapsulates the business logic in an app; it is the only portion of the UI that uses the service layer.

For us `service layer` are context properties from Nim. When we encapsulate that, then no matter how backend design will change we can replace it with different implementation, potentially without touching store's interface at all (but ofc probably some changes will be needed to simplify and reorganize stuff).

> A store can have child stores which can be further forwarded to sub-trees in the UI.

For that reason from UI part we should instantiate only single `RootStore` directly. Other stores should be substores, instantiated inside each other, creating tree-like hierarchy.

> Ideally, a store has an interface that defines the API for the store to be tested. Views would only see this interface so that they do not depend on a concrete store. The store that inherits the abstract interface then fills it with values from the required service and feeds the UI. Alternatively, developers can also use the store interface and feed the UI via static simulation data or an automated simulation backend that runs the states required to provide the desired data.

We have testability/simulation solved differently (via overriding import paths and providing stubs). Imo it's ok because qml doesn't support interfaces in a reliable manner anyway. Thanks to that we can always use concrete types.

So my proposition is to just start following those basic rules. Besides that, as described earlier, I opt for keeping the flow unidirectional - state is exposed via read-only properties and changes are done methods.

Reply:
> ... in the future the idea is to get rid of context properties (currently are there because of nim) and have QML types instead. At least that's the info I have of how the Qt/C++ layer will be implemented (@stefandunca @saledjenic please correct me if I'm wrong).

@alexandraB99 indeed you are right that the C++ plan promises to deliver the right abstractions. However, given that there is no concrete timeline for the C++ refactoring I agree with @micieslak's proposal not to wait for the C++ to start fixing "broken windows".

It looks to me that @micieslak gives a good account of why we need to abstract the current context properties to lower the coupling between QMl and NIM. This will serve future C++ work already (indirectly by providing guidelines). The proposed solutions (hide context properties under Store's interface functions and properties as generic interface) are generally good approaches, that have to be weighed on a case-by-case basis.

> That said, all these xxxInst properties are there to prepare for the upcoming QML types, will look like something like here for example if you see
>  property FilterAndBrowseModel searchAndBrowseModel: FilterAndBrowseModel {
>          id: searchBrowseModel

An equivalent of the above in this case would be

```qml
// RootStore.qml
property GlobalUtils globalUtils: GlobalUtils {}

// GlobalUtils.qml
QtObject {
  // From @Michal's answer
  readonly property bool showBrowserSelector: localAccountSensitiveSettings.showBrowserSelector
  readonly property bool openLinksInStatus: localAccountSensitiveSettings.openLinksInStatus
  function copyToClipboard(text) { globalUtils.copyToClipboard(text) }
  function plainText(text) { globalUtils.plainText(text) }
}

```

Reply:
I agree on this approach. We should try to remove all dependencies by creating just only the explicit needed interfaces in stores. So, in this particular case I will opt on removing the `globalUtilsInst` property and expose only the needed and specific stuff.

Reply:
@stefandunca exactly the same thing I'm saying, so we'd still need the properties to instantiate the objects and refer to those, to my understanding @micieslak is proposing to delete `property var globalUtilsInt: globalUtils` which in the future will be alterated to `property GlobalUtils globalUtils: GlobalUtils {}` and just call directly the backend object in wrapper functions and properties like `function plainText(text) { globalUtils.plainText(text) }` where globalUtils in this case is the context property. I am not saying I am against wrapping all necessary stuff into functions and properties in the store and expose only these, this is anyway something we had in mind to do as a second step in refactoring process I am just concerned on calling backend objects directly in wrapper functions and properties declared in stores.

Reply:
> I am just concerned on calling backend objects directly in wrapper functions and properties declared in stores.

What we want to do is totally opposite, we reduce directness and introduce abstraction layer. Access to context property is hidden behind that abstraction. Not used directly from UI. So it's pointless to expose it now only to revert it back later. And effectively we are introducing less changes bc it's no need to refactore e.g:
```
function copyToClipboard(text) {
    globalUtils.copyToClipboard(text)
}
```
into
```
function copyToClipboard(text) {
    globalUtilsInst.copyToClipboard(text)
}
```

Reply:
One direction might be worth investigating is using nim defined qml types. 
From what I see nimqml supports qmlRegisterSingletonType and qmlRegisterType. I've done this quick test with qmlRegisterSingletonType for globalUtils and it can be registered as qml singleton type:

```
#AppController.nim
proc startupDidLoad*(self: AppController) = 
...
  let id = qmlRegisterSingletonType("StatusBackend", 1, 0, "Utils", proc(): Utils = singletonInstance.utils)
  singletonInstance.engine.load(newQUrl("qrc:///main.qml"))
...
```

```
Global.qml
pragma Singleton

import QtQml 2.14
import StatusBackend 1.0 as Backend
...
    function openLink(link) {
        // Qt sometimes inserts random HTML tags; and this will break on invalid URL inside QDesktopServices::openUrl(link)
        link = Backend.Utils.plainText(link);
        if (localAccountSensitiveSettings.showBrowserSelector) {
            openChooseBrowserPopup(link);
        } else {
            if (localAccountSensitiveSettings.openLinksInStatus) {
                changeAppSectionBySectionType(Constants.appSection.browser);
                openLinkInBrowser(link);
            } else {
                Qt.openUrlExternally(link);
            }
        }
    }
```

This will help in setting some boundaries in the app and slowly removing context properties. Not sure it works well with qmlRegisterType also, but for singletons it looks promising.

Reply:
I'm afraid that such singletons would hinder testability a lot. UI singetons like `Global.qml` or `Utils.qml` really should be stateless and not relay on backend. Backend access should be hidden behind store's API.

Reply:
@alexandraB99, to be clear I understand that exposing things in this way `property var globalUtilsInt: globalUtils` is tempting bc we may expect similar objects to be exposed as a part of stores' API in the future. However, in my opinion, the disadvantages outweigh the advantages here, main of them:
1) Stores API is totally vague, we don't know anything about the the type of given context property and about types/names/models inside. We are forced to dig into nim code then. And all views rely on that unclear API right now.
2). We tend to take those context props exposed from stores and pass directly to subviews as `var` properties, instead of passing typed stores
3) There are low chances that new api will reflect current structure of context props. We shouldn't expect from backend developers to adhere to that structure. Therefore introducing new backend will affect many parts of the app from the very beginning, including views. Having stores api properly defined, we could in the first step replace backend potentially even without touching views.

If we need group some props/methods we can create substore for that or other type defined in store's namespace. In our case we could even define a separate type especially for wrapping given context property:

```
///AppLayouts/stores/GlobalUtils.qml

GlobalUtils {
   // some strongly typed, readonly props here
   function plainText(text) {
      return globalUtils.plainText(text)
   }
   // other methods here
}
```
and then use in our `AppLayouts/stores/RootStore`:
```
readonly property GlobalUtils globalUtils: GlobalUtils {}
```
It's the same (in terms of exposed structure) but we gain clear, strongly typed interface, including support from QtCreator. We need to bother about nim code only when implementing stores and store-related wrapper types like that `GlobalUtils`. But no longer in any other components like views.

Reply:
> One direction might be worth investigating is using nim defined qml types. From what I see nimqml supports qmlRegisterSingletonType and qmlRegisterType. ...

@alexjba Great finding!

> @alexjba I'm afraid that such singletons would hinder testability a lot. ...

@micieslak I see @alexjba's post addressing the problem of "context properties" not necessarily abstracting NIM implementation. It looks like a contender for a quick fix to the current way of exposing NIM QObjects as context properties.

In theory, abusing context properties slows down the QML execution due to the slow look-up of properties in the global context. That's it, referenced NIM QObjects are searched up the QML engine's object tree and always found at the root of the tree. Additionally, context properties break the code clarity by using magic properties defined outside the QML file. 

Exposing NIM QObjects as singletons might help speeding up the QML code while providing a cleaner way to expose NIM QObjects in QML by using clear imports instead of magic variables.

Reply:
Yeah, from that perspective (context properties -> Singletons) indeed it may improve performance. But still they shouldn't be accessed in qml's singetons like in provided example (`openLink` defined in `Global.qml` singleton), bc then it would be similarly problematic for testing as it's now. But yeah, hiding access to nim part in stores + converting ctx props to Singletons sounds like good combination.

Reply:
Ohh, yes. Please consider the code snippet just as an easy test example of singleton qml type registration. Wasn't my intention to propose a function openLink(link) implementation.

I agree on the proposed Global.qml and stores reconfiguration and since we're doing this I think it's a good time to also reconsider the way we integrate nim and qml. Using nim defined qml singleton types looks like a promising replacement of root context properties and it will help us better define and contain the nim - qml integration layer.

Reply:
@micieslak thanks for bringing up the discussion.

> We should eliminate exposing injected context properties from stores because it makes stores API really unclear.

I agree. According to qml core architecture "stores" purpose is to abstract away all backend. Exposing anything nim related (*moduleInst, globalUtils, etc.) as a public API is just a violation of this principle.

> @micieslak I see @alexjba's post addressing the problem of "context properties" not necessarily abstracting NIM implementation. It looks like a contender for a quick fix to the current way of exposing NIM QObjects as context properties.

@stefandunca well said. I just hope it will be less tempting to do `import StatusBackend` than using magic context properties outside the stores. We have the rule to not use `*moduleIns` outside of stores but it was never really applied. I think it is easier to enforce `import StatusBackend` to be used only in stores, we could even come up with some CI checker.

> But yeah, hiding access to nim part in stores + converting ctx props to Singletons sounds like good combination.

:+1:  Let's finally make "stores" fulfill its purpose :pray: 
- encapsulate everything nim related into store, do not expose any *moduleInst, globalUtilsInst as public API
- use Singletons instead of context properties
- enforce `import StatusBackend` to be used only in Stores